### DISCUSSION TITLE
LLM Tracing UI - Tags

### DISCUSSION THREAD

**Discussion by the-sean-c** [mlflow/mlflow#12927] [Q&A] [answered]
I'm working with Langchain, and I'm using the MLFlow LLM features to autolog traces. In the UI, on the "Runs" tab I can see tags that are set with`mlflow.set_tag("tag_key", "tag_value")`, but I can't see them on the "Traces" tab. I can manually add tags here one-by-one, and these seem to persist.

Are "Traces" tags set and stored differently to "Run" tags? If so, how? (I can't seem to find it in the docs or the discussion).

LLM tracking is such an awesome addition to MLFlow by the way. It was a natural addition to the platform and I knew you guys would nail it!

**Comment by B-Step62** [comment, marked as answer]
@the-sean-c Run tags and Traces tags are independent - similar to how Experiment tags are different as well.

There is `set_trace_tag()` API in [MlflowClient](https://github.com/mlflow/mlflow/blob/786a95d8f2e50f8bff6e2461feb7df6b631ec6fb/mlflow/tracking/client.py#L983) and it is possible to use it for setting tags during Langchain autologging.
```
mlflow.langchain.autolog()
client = mlflow.MlflowClient()

with mlflow.start_run():
    response = chain.invoke()

    # Set a tag to the trace just generated
    trace = mlflow.get_last_active_trace()
    client.set_trace_tag(trace.request_id, "key", "value")
```

However, it may not be very convenient because you need to get the ID from the trace like this. Is your tag key or value dynamic, or fixed when calling `mlflow.langchain.autolog()`? If it is fixed, we can consider adding a parameter to `mlflow.langchain.autolog()` so the traces will have that tags automatically.

**Reply by the-sean-c** [reply]
Gotcha! It would be awesome to have this work like `mlflow.set_tag("tag_key", "tag_value")` works for autologging, but I can make it work the way you have it laid out above.

With tag keys, whether they're fixed or dynamic depends on the task. Often they're fixed, often they're dynamic.

Thanks for the response!

**Comment by B-Step62** [comment]
FYI: In the next MLflow release, we are going to ship the new `update_current_trace` API, with which you can dynamically set tags for traces during autologging. 
https://github.com/mlflow/mlflow/blob/d8732dd8314268306c10c49137cf5a1c18e1c633/mlflow/tracing/fluent.py#L574

**Reply by the-sean-c** [reply]
Awesome! Can't wait to try it out!

**Reply by mattijsdp** [reply]
Is that actually possible? Does the autologged trace not end before the invoke model (or similar) returns?
